core/consensus/babe/impl/babe_synchronizer_impl.cpp

```diff
@@ -3,138 +3,858 @@
+                     header.number,
+                     block.hash.toHex(),
+                     peer_id.toBase58(),
+                     last_finalized_block.number);
+            continue;
+          }
+
+          SL_TRACE(self->log_,
+                   "Skip block #{} hash={} received from {}: "
+                   "it is below the last finalized (#{})",
+                   header.number,
+                   block.hash.toHex(),
+                   peer_id.toBase58(),
+                   last_finalized_block.number);
+          continue;
+        }
+
+        // Check if block is not discarded
+        if (last_finalized_block.number + 1 == header.number) {
+          if (last_finalized_block.hash != header.parent_hash) {
+            SL_ERROR(self->log_,
+                     "Can't complete blocks loading from {} starting from "
+                     "block #{} hash={}: "
+                     "Received discarded block (#{} hash={})",
+                     peer_id.toBase58(),
+                     from.number,
+                     from.hash.toHex(),
+                     header.number,
+                     header.parent_hash.toHex());
+            handler(Error::DISCARDED_BLOCK);
+            return;
+          }
+
+          // Start to check parents
+          parent_hash = header.parent_hash;
+        }
+
+        // Check if block is in chain
+        static const primitives::BlockHash zero_hash;
+        if (parent_hash != header.parent_hash && parent_hash != zero_hash) {
+          SL_ERROR(self->log_,
+                   "Can't complete blocks loading from {} starting from "
+                   "block #{} hash={}: "
+                   "Received block is not descendant of previous",
+                   peer_id.toBase58(),
+                   from.number,
+                   from.hash.toHex());
+          handler(Error::WRONG_ORDER);
+          return;
+        }
+
+        // Check if hash is valid
+        auto calculated_hash =
+            self->hasher_->blake2b_256(scale::encode(header).value());
+        if (block.hash != calculated_hash) {
+          SL_ERROR(self->log_,
+                   "Can't complete blocks loading from {} starting from "
+                   "block #{} hash={}: "
+                   "Received block whose hash does not match the header",
+                   peer_id.toBase58(),
+                   from.number,
+                   from.hash.toHex());
+          handler(Error::INVALID_HASH);
+          return;
+        }
+
+        last_loaded_block.number = header.number;
+        last_loaded_block.hash = block.hash;
+
+        parent_hash = block.hash;
+
+        // Add block in queue and save peer or just add peer for existing record
+        auto it = self->known_blocks_.find(block.hash);
+        if (it == self->known_blocks_.end()) {
+          self->known_blocks_.emplace(block.hash, KnownBlock{block, {peer_id}});
+        } else {
+          it->second.peers.emplace(peer_id);
```

Comment:
why do you need list of peers for every block in sync?

Reply:
They are using to resume of loading blocks. First available peer would be enough for that.